File: mobile/lib/model/alert.dart
import 'package:emergency_pulse/model/alert_action.dart';
import 'package:emergency_pulse/model/alert_type.dart';

class AlertModel {
  final String alertHashId;
  final AlertType alertType;
  final String imei;
  final String name;
  final String address;
  final String contactNo;
  final String lat;
  final String lng;
  final String accuracyMeters;
  final String notes;
  final String deviceModel;
  final String deviceBrand;
  final String deviceVersion;
  final String deviceName;
  final String deviceBatteryLevel;
  final String distance;
  final String createdAt;

  AlertModel({
    required this.alertHashId,
    required this.alertType,
    required this.imei,
    required this.name,
    required this.address,
    required this.contactNo,
    required this.lat,
    required this.lng,
    required this.accuracyMeters,
    required this.notes,
    required this.deviceModel,
    required this.deviceBrand,
    required this.deviceVersion,
    required this.deviceName,
    required this.deviceBatteryLevel,
    required this.distance,
    required this.createdAt,
  });

  factory AlertModel.fromJson(Map<String, dynamic> json) {
    return AlertModel(
      alertHashId: json['alertHashId'],
      alertType: AlertType.values.firstWhere(
        (e) => e.name == json['alertType'],
      ),
      imei: json['imei'],
      name: json['name'],
      address: json['address'],
      contactNo: json['contactNo'],
      lat: json['lat'],
      lng: json['lng'],
      accuracyMeters: json['accuracyMeters'],
      notes: json['notes'],
      deviceModel: json['deviceModel'],
      deviceBrand: json['deviceBrand'],
      deviceVersion: json['deviceVersion'],
      deviceName: json['deviceName'],
      deviceBatteryLevel: json['deviceBatteryLevel'],
      distance: json['distance'].toString(),
      createdAt: json['createdAt'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'alertHashId': alertHashId,
      'alertType': alertType.name,
      'imei': imei,
      'name': name,
      'address': address,
      'contactNo': contactNo,
      'lat': lat,
      'lng': lng,
      'accuracyMeters': accuracyMeters,
      'notes': notes,
      'deviceModel': deviceModel,
      'deviceBrand': deviceBrand,
      'deviceVersion': deviceVersion,
      'deviceName': deviceName,
      'deviceBatteryLevel': deviceBatteryLevel,
      'distance': distance,
      'createdAt': createdAt,
    };
  }
}

File: backend/internal/model/alert.go
package model

import (
	"log/slog"
	"pulse/internal/dto"
	"pulse/internal/hash"
	"pulse/internal/utils"
	"time"
)

type AlertModel struct {
	AlertID            int32     `db:"alert_id"`
	AlertType          string    `db:"alert_type"`
	Imei               string    `db:"imei"`
	Name               *string   `db:"name"`
	Address            *string   `db:"address"`
	ContactNo          *string   `db:"contact_no"`
	Lat                string    `db:"lat"`
	Lng                string    `db:"lng"`
	AccuracyMeters     string    `db:"accuracy_meters"`
	DeviceBrand        string    `db:"device_brand"`
	DeviceModel        string    `db:"device_model"`
	DeviceVersion      string    `db:"device_version"`
	DeviceName         string    `db:"device_name"`
	DeviceBatteryLevel string    `db:"device_battery_level"`
	Notes              *string   `db:"notes"`
	Distance           float64   `db:"distance"`
	ResponderCount     int       `db:"responder_count"`
	Action             string    `db:"action"`
	ActionAt           time.Time `db:"action_at"`
	CreatedAt          time.Time `db:"created_at"`
}

func (a AlertModel) ToDTO() (dto.AlertDTO, error) {
	hashID, err := hash.HashAlertID(int(a.AlertID))

	if err != nil {
		slog.Error("[AlertModel.ToDTO] [1] ERROR: " + err.Error())
		return dto.AlertDTO{}, err
	}

	return dto.AlertDTO{
		AlertHashID:        hashID,
		AlertType:          a.AlertType,
		Imei:               a.Imei,
		Name:               a.Name,
		Address:            a.Address,
		ContactNo:          a.ContactNo,
		Lat:                a.Lat,
		Lng:                a.Lng,
		AccuracyMeters:     a.AccuracyMeters,
		Distance:           a.Distance,
		DeviceBrand:        a.DeviceBrand,
		DeviceModel:        a.DeviceModel,
		DeviceVersion:      a.DeviceVersion,
		DeviceName:         a.DeviceName,
		DeviceBatteryLevel: a.DeviceBatteryLevel,
		Notes:              a.Notes,
		ResponderCount:     a.ResponderCount,
		Action:             a.Action,
		ActionAt:           utils.TimeToISO8601(a.ActionAt),
		CreatedAt:          utils.TimeToISO8601(a.CreatedAt),
	}, nil
}
